language/move-prover/bytecode/src/mono_analysis.rs

```diff
@@ -42,29 +58,96 @@ impl MonoAnalysisProcessor {
+        // `P[TypeCtor<u64, A>] && P[TypeCtor<address, B>]`.
+        let code = std::mem::take(&mut data.code);
+        let mut rewriter = TypeQuantRewriter {
+            builder: FunctionDataBuilder::new(fun_env, data),
+            instantiated_memory: Default::default(),
+            quants_eliminated: false,
         };
-        analyzer.analyze();
-        let Analyzer { info, .. } = analyzer;
-        env.set_extension(info);
-    }
-
-    fn is_single_run(&self) -> bool {
-        true
+        for bc in code {
+            if let Bytecode::Prop(id, kind, exp) = bc {
+                let exp = rewriter.rewrite_type_quant(exp);
+                rewriter.builder.emit(Bytecode::Prop(id, kind, exp));
+            } else {
+                rewriter.builder.emit(bc);
+            }
+        }
+        if rewriter.quants_eliminated {
+            // We need to go a second time over the code to expand `SaveMem` instructions.
+            // Where we previous had `SaveMem<R<type_var>>` we need to create one SaveMem
+            // for every instance.
+            let code = std::mem::take(&mut rewriter.builder.data.code);
+            for bc in code {
+                match bc {
+                    Bytecode::SaveMem(id, label, mem) => {
+                        if rewriter.instantiated_memory.contains_key(&label) {
+                            for inst in rewriter.instantiated_memory.get(&label).unwrap() {
+                                rewriter.builder.emit(Bytecode::SaveMem(
+                                    id,
+                                    label,
+                                    inst.to_owned(),
+                                ));
+                            }
+                        } else if !mem
+                            .inst
+                            .iter()
+                            .any(|ty| ty.contains(&|t| matches!(t, Type::TypeLocal(_))))
+                        {
+                            // Only retain the SaveMem if it does not contain type locals.
+                            // Such SaveMem's can result from zero expansions during quantifier
+                            // elimination, and they are dead.
+                            rewriter.builder.emit(SaveMem(id, label, mem));
+                        }
+                    }
+                    _ => rewriter.builder.emit(bc),
+                }
+            }
+            rewriter
+                .builder
+                .global_env()
+                .set_extension(RunAnalysisAgain());
+        }
+        rewriter.builder.data
     }
 
     fn name(&self) -> String {
         "mono_analysis".to_owned()
     }
 
+    fn initialize(&self, env: &GlobalEnv, targets: &mut FunctionTargetsHolder) {
+        self.analyze(env, targets);
+    }
+
+    fn finalize(&self, env: &GlobalEnv, targets: &mut FunctionTargetsHolder) {
+        if env.has_extension::<RunAnalysisAgain>() {
```

Comment:
It feels to me a bit brutal to pass the `RunAnalysisAgain` mark around with `env.has_extension()`. Can we confine it in the `MonoAnalysisProcessor::analyze()` function?

Reply:
No we can't. This is a generic problem right now: we cannot persist any state between multiple runs of process (they are run for each function) unless (a) via env extension or (b) function target annotations and summaries.

This is something we should to the framework, I needed it already multiple times before.

@sblackshear

Reply:
FWIW, I think the solution for this is to actually registerer a *factory* for a FunctionTargetProcessor. Then when a processor is needed, it can be created with the factory, and can have instance state.